rust/pact_verifier/src/pact_broker.rs

```diff
@@ -408,37 +421,44 @@ pub fn fetch_pacts_from_broker(
                             provider_name, broker_url)),
                 _ => err
             }
+        })?;
+
+    let pact_links = hal_client.iter_links(s!("pacts"))?;
+
+    let results: Vec<_> = futures::stream::iter(pact_links)
+        .map(|pact_link| {
+            match pact_link.clone().href {
+                Some(_) => Ok((hal_client.clone(), pact_link)),
+                None => Err(
+                    PactBrokerError::LinkError(
+                        format!(
+                            "Expected a HAL+JSON response from the pact broker, but got a link with no HREF. URL: '{}', LINK: '{:?}'",
+                            &hal_client.url,
+                            pact_link
+                        )
+                    )
+                )
+            }
         })
-        .and_then(|hal_client| {
-            hal_client.iter_links(s!("pacts"))
-                .map(|pact_links| (hal_client, pact_links))
+        .and_then(|(hal_client, pact_link)| async {
+            let pact_json = hal_client.fetch_url(
+                pact_link.clone(),
+                template_values.clone()
+            ).await?;
+            Ok((pact_link, pact_json))
         })
-        .and_then(move |(hal_client, pact_links)| {
-            let client_url_cloned = hal_client.url.clone();
-
-            futures::stream::iter_ok::<_, PactBrokerError>(pact_links)
-                .and_then(move |pact_link| {
-                    match pact_link.clone().href {
-                        Some(_) => Ok(pact_link),
-                        None => Err(
-                            PactBrokerError::LinkError(format!("Expected a HAL+JSON response from the pact broker, but got a link with no HREF. URL: '{}', LINK: '{:?}'",
-                                client_url_cloned, pact_link))
-                        )
-                    }
-                })
-                .and_then(move |pact_link| {
-                  hal_client.clone().fetch_url(&pact_link, template_values.clone())
-                    .map(move |pact_json| {
-                      let pact = Pact::from_json(&pact_link.href.clone().unwrap(), &pact_json);
-                      let links = links_from_json(&pact_json);
-                      (pact, links)
-                    })
-                })
-                .then(|result| {
-                    Ok(result)
-                })
-                .collect()
+        .map_ok(|(pact_link, pact_json)| {
+            let href = pact_link.href.unwrap();
+            let pact = Pact::from_json(&href, &pact_json);
+            let links = links_from_json(&pact_json);
+
+            (pact, links)
         })
+        .into_stream()
```

Comment:
Can very easily refactor download of several pact files to be concurrent instead of sequential, if there's any value to that.